docs/man/btdownloadheadless.bittornado.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.33.
.TH BTDOWNLOADHEADLESS "1" "August 2004" "btdownloadheadless (bittornado)" "User Commands"
.SH NAME
btdownloadheadless \- headless bittornado download interface
.SH DESCRIPTION
Downloads torrents in the same manner as btdownloadcurses, but does not require
a terminal to stay alive.
.SH OPTIONS
This program follows the usual GNU command-line syntax, with long options
starting with two dashes ('-'). A summary of options is included below.

.PP
\fB\-\-max_uploads\fR \fInumber\fP
.IP
the maximum \fInumber\fP of uploads to allow at once. (defaults to 7)
.PP
\fB\-\-keepalive_interval\fR \fIseconds\fP
.IP
number of \fIseconds\fP to pause between sending keepalives (defaults to
120.0)
.PP
\fB\-\-download_slice_size\fR \fIbytes\fP
.IP
How many \fIbytes\fP to query for per request. (defaults to 16384)
.PP
\fB\-\-upload_unit_size\fR \fIbytes\fP
.IP
when limiting upload rate, how many \fIbytes\fP to send at a time (defaults
to 1460)
.PP
\fB\-\-request_backlog\fR \fInumber\fP
.IP
maximum \fInumber\fP of requests to keep in a single pipe at once.
(defaults to 10)
.PP
\fB\-\-max_message_length\fR \fIlength\fP
.IP
maximum \fIlength\fP prefix encoding you'll accept over the wire - larger
values get the connection dropped. (defaults to 8388608)
.PP
\fB\-\-ip\fR \fIip\fP
.IP
\fIip\fP to report you have to the tracker. (defaults to '')
.PP
\fB\-\-minport\fR \fIport\fP
.IP
minimum \fIport\fP to listen on, counts up if unavailable (defaults to
10000)
.PP
\fB\-\-maxport\fR \fIport\fP
.IP
maximum \fIport\fP to listen on (defaults to 60000)
.PP
\fB\-\-random_port\fR \fI 0 | 1 \fP
.IP
whether to choose randomly inside the port range instead of counting
up linearly (defaults to 1)
.PP
\fB\-\-responsefile\fR \fIfile\fP
.IP
\fIfile\fP the server response was stored in, alternative to url (defaults
to '')
.PP
\fB\-\-url\fR \fIURL\fP
.IP
\fIURL\fP to get file from, alternative to responsefile (defaults to '')
.PP
\fB\-\-selector_enabled\fR \fI 0 | 1 \fP
.IP
whether to enable the file selector and fast resume function
(defaults to 1)
.PP
\fB\-\-expire_cache_data\fR \fIdays\fP
.IP
the number of \fIdays\fP after which you wish to expire old cache data (0 =
disabled) (defaults to 10)
.PP
\fB\-\-priority\fR \fI -1|0|1|2[,-1|0|1|2] \fP
.IP
a list of file priorities separated by commas, must be one per file, 0 = highest, 1 =
normal, 2 = lowest, \fB\-1\fR = download disabled (defaults to ''). Order is based
on the file/torrent order as shown by btshowmetainfo. For example, to download only
the third of four files use: --priority -1,-1,2,-1
.PP
\fB\-\-priority\fR \fI -1|0|1|2[,-1|0|1|2] \fP
.IP
a list of file priorities separated by commas, must be one per file, 0 = highest, 1 =
normal, 2 = lowest, \fB\-1\fR = download disabled (defaults to ''). Order is based
on the file/torrent order as shown by btshowmetainfo. For example, to download only
the third of four files use: --priority -1,-1,2,-1
.PP
\fB\-\-saveas\fR \fIfilename\fP
.IP
local \fIfilename\fP to save the file as, null indicates query user (defaults to '')
.PP
\fB\-\-timeout\fR \fIseconds\fP
.IP
time in \fIseconds\fP to wait between closing sockets which nothing has been received on (defaults to
300.0)
.PP
\fB\-\-timeout_check_interval\fR \fIseconds\fP
.IP
time to wait in \fIseconds\fP between checking if any connections have timed out (defaults to 60.0)
.PP
\fB\-\-max_slice_length\fR \fIlength\fP
.IP
maximum \fIlength\fP slice to send to peers, larger requests are ignored (defaults to 131072)
.PP
\fB\-\-max_rate_period\fR \fIseconds\fP
.IP
maximum amount of time in \fIseconds\fP to guess the current rate estimate represents (defaults to 20.0)
.PP
\fB\-\-bind\fR \fIip[,hostname]\fP
.IP
comma-separated list of \fIips/hostnames\fP to bind to locally (defaults to '')
.PP
\fB\-\-ipv6_enabled\fR \fI 0 | 1 \fP
.IP
allow the client to connect to peers via IPv6 (defaults to 0)
.PP
\fB\-\-ipv6_binds_v4\fR \fI 0 | 1 \fP
.IP
set if an IPv6 server socket will also field IPv4 connections (defaults to 1)
.PP
\fB\-\-upnp_nat_access\fR \fI 0 | 1 | 2 \fP
.IP
attempt to autoconfigure a UPnP router to forward a server port (0 = disabled, 1 = mode 1
[fast], 2 = mode 2 [slow]) (defaults to 1)
.PP
\fB\-\-upload_rate_fudge\fR \fIseconds\fP
.IP
time equivalent in \fIseconds\fP of writing to kernel-level TCP buffer, for rate adjustment (defaults to
5.0)
.PP
\fB\-\-tcp_ack_fudge\fR \fIoverhead\fP
.IP
how much TCP ACK download \fIoverhead\fP to add to upload rate calculations (0 = disabled)
(defaults to 0.029999999999999999)
.PP
\fB\-\-display_interval\fR \fIseconds\fP
.IP
time in \fIseconds\fP between updates of displayed information (defaults to 0.5)
.PP
\fB\-\-rerequest_interval\fR \fIseconds\fP
.IP
time to wait, in \fIseconds\fP, between requesting more peers (defaults to 300)
.PP
\fB\-\-min_peers\fR \fInumber\fP
.IP
minimum \fInumber\fP of peers to not do rerequesting (defaults to 20)
.PP
\fB\-\-http_timeout\fR \fIseconds\fP
.IP
number of \fIseconds\fP to wait before assuming that an http connection has timed out (defaults
to 60)
.PP
\fB\-\-max_initiate\fR \fInumber\fP
.IP
\fInumber\fP of peers at which to stop initiating new connections (defaults to 40)
.PP
\fB\-\-check_hashes\fR \fI 0 | 1 \fP
.IP
whether to check hashes on disk (defaults to 1)
.PP
\fB\-\-max_upload_rate\fR \fIkB/s\fP
.IP
maximum \fIkB/s\fP to upload at (0 = no limit, \fB\-1\fR = automatic) (defaults to 0)
.PP
\fB\-\-max_download_rate\fR \fIkB/s\fP
.IP
maximum \fIkB/s\fP to download at (0 = no limit) (defaults to 0)
.PP
\fB\-\-alloc_type\fR \fInormal | background | pre-allocate | sparse\fP
.IP
allocation type (may be normal, background, pre-allocate or sparse) (defaults to
\&'normal')
.PP
\fB\-\-alloc_rate\fR \fIMiB/s\fP
.IP
rate (in \fIMiB/s\fP) to allocate space at using background allocation (defaults to 2.0)
.PP
\fB\-\-buffer_reads\fR \fI 0 | 1 \fP
.IP
whether to buffer disk reads (defaults to 1)
.PP
\fB\-\-write_buffer_size\fR \fIspace\fP
.IP
the maximum amount of \fIspace\fP to use for buffering disk writes (in megabytes, 0 = disabled)
(defaults to 4)
.PP
\fB\-\-snub_time\fR \fIseconds\fP
.IP
\fIseconds\fP to wait for data to come in over a connection before assuming it's
semi-permanently choked (defaults to 30.0)
.PP
\fB\-\-spew\fR \fI 0 | 1 \fP
.IP
whether to display diagnostic info to stdout (defaults to 0)
.PP
\fB\-\-rarest_first_cutoff\fR \fInumber\fP
.IP
\fInumber\fP of downloads at which to switch from random to rarest first (defaults to 2)
.PP
\fB\-\-rarest_first_priority_cutoff\fR \fInumber\fP
.IP
the \fInumber\fP of peers which need to have a piece before other partials take priority over
rarest first (defaults to 5)
.PP
\fB\-\-min_uploads\fR \fInumber\fP
.IP
the \fInumber\fP of uploads to fill out to with extra optimistic unchokes (defaults to 4)
.PP
\fB\-\-max_files_open\fR \fInumber\fP
.IP
the maximum \fInumber\fP of files to keep open at a time, 0 means no limit (defaults to 50)
.PP
\fB\-\-round_robin_period\fR \fIseconds\fP
.IP
the number of \fIseconds\fP between the client's switching upload targets (defaults to 30)
.PP
\fB\-\-super_seeder\fR \fI 0 | 1 \fP
.IP
whether to use special upload-efficiency-maximizing routines (only for dedicated seeds)
(defaults to 0)
.PP
\fB\-\-security\fR \fI 0 | 1 \fP
.IP
whether to enable extra security features intended to prevent abuse (defaults to 1)
.PP
\fB\-\-max_connections\fR \fInumber\fP
.IP
the absolute maximum \fInumber\fP of peers to connect with (0 = no limit) (defaults to 0)
.PP
\fB\-\-auto_kick\fR \fI 0 | 1 \fP
.IP
whether to allow the client to automatically kick/ban peers that send bad data (defaults
to 1)
.PP
\fB\-\-double_check\fR \fI 0 | 1 \fP
.IP
whether to double-check data being written to the disk for errors (may increase CPU load)
(defaults to 1)
.PP
\fB\-\-triple_check\fR \fI 0 | 1 \fP
.IP
whether to thoroughly check data being written to the disk (may slow disk access)
(defaults to 0)
.PP
\fB\-\-lock_files\fR \fI 0 | 1 \fP
.IP
whether to lock files the client is working with (defaults to 1)
.PP
\fB\-\-lock_while_reading\fR \fI 0 | 1 \fP
.IP
whether to lock access to files being read (defaults to 0)
.PP
\fB\-\-auto_flush\fR \fIminutes\fP
.IP
\fIminutes\fP between automatic flushes to disk (0 = disabled) (defaults to 0)
.PP
\fB\-\-save_options\fR <arg>
.IP
whether to save the current options as the new default configuration
(only for btdownloadheadless.py) (defaults to 0)
.PP
.SH SEE ALSO
.BR bittorrent-downloader (1),
.BR btmakemetafile (1).
.br
.SH AUTHOR
This manual page was written by Micah Anderson <micah@riseup.net>,
for the Debian GNU/Linux system (but may be used by others).


btdownloadheadless.py
#!/usr/bin/env python

# Written by Bram Cohen
# see LICENSE.txt for license information

import sys
import os
import time
import random
import socket
import hashlib
import threading
from BitTornado.Client.download_bt1 import BT1Download, defaults, \
    parse_params, get_usage, get_response
from BitTornado.Network.RawServer import RawServer
from BitTornado.Network.SocketHandler import UPnP_ERROR
from BitTornado.Meta.bencode import bencode
from BitTornado.Network.natpunch import UPnP_test
from BitTornado.clock import clock
from BitTornado import version
from BitTornado.Application.ConfigDir import ConfigDir
from BitTornado.Application.PeerID import createPeerID

PROFILER = False


def hours(n):
    if n == 0:
        return 'complete!'
    try:
        n = int(n)
        assert n >= 0 and n < 5184000  # 60 days
    except (AssertionError, ValueError):
        return '<unknown>'
    m, s = divmod(n, 60)
    h, m = divmod(m, 60)
    if h > 0:
        return '%d hour %02d min %02d sec' % (h, m, s)
    else:
        return '%d min %02d sec' % (m, s)


class HeadlessDisplayer:
    def __init__(self):
        self.done = False
        self.file = ''
        self.percentDone = ''
        self.timeEst = ''
        self.downloadTo = ''
        self.downRate = ''
        self.upRate = ''
        self.shareRating = ''
        self.seedStatus = ''
        self.peerStatus = ''
        self.errors = []
        self.last_update_time = -1

    def finished(self):
        self.done = True
        self.percentDone = '100'
        self.timeEst = 'Download Succeeded!'
        self.downRate = ''
        self.display()

    def failed(self):
        self.done = True
        self.percentDone = '0'
        self.timeEst = 'Download Failed!'
        self.downRate = ''
        self.display()

    def error(self, errormsg):
        self.errors.append(errormsg)
        self.display()

    def display(self, dpflag=threading.Event(), fractionDone=None,
                timeEst=None, downRate=None, upRate=None, activity=None,
                statistics=None, **kws):
        if self.last_update_time + 0.1 > clock() and \
                fractionDone not in (0.0, 1.0) and activity is not None:
            return
        self.last_update_time = clock()
        if fractionDone is not None:
            self.percentDone = str(float(int(fractionDone * 1000)) / 10)
        if timeEst is not None:
            self.timeEst = hours(timeEst)
        if activity is not None and not self.done:
            self.timeEst = activity
        if downRate is not None:
            self.downRate = '%.1f kB/s' % (float(downRate) / (1 << 10))
        if upRate is not None:
            self.upRate = '%.1f kB/s' % (float(upRate) / (1 << 10))
        if statistics is not None:
            if statistics.shareRating < 0 or statistics.shareRating > 100:
                self.shareRating = 'oo  ({:.1f} MB up / {:.1f} MB down)' \
                    ''.format(float(statistics.upTotal) / (1 << 20),
                              float(statistics.downTotal) / (1 << 20))
            else:
                self.shareRating = '{:.3f}  ({:.1f} MB up / {:.1f} MB down)' \
                    ''.format(statistics.shareRating,
                              float(statistics.upTotal) / (1 << 20),
                              float(statistics.downTotal) / (1 << 20))
            if not self.done:
                self.seedStatus = '{:d} seen now, plus {:.3f} distributed ' \
                    'copies'.format(statistics.numSeeds,
                                    statistics.numCopies)
            else:
                self.seedStatus = '{:d} seen recently, plus {:.3f} ' \
                    'distributed copies'.format(statistics.numOldSeeds,
                                                statistics.numCopies)
            self.peerStatus = '{:d} seen now, {:.1%} done at {:.1f} kB/s' \
                ''.format(statistics.numPeers,
                          statistics.percentDone / 100,
                          float(statistics.torrentRate) / (1 << 10))
        print '\n\n\n\n'
        for err in self.errors:
            print 'ERROR:\n' + err + '\n'
        print 'saving:        ', self.file
        print 'percent done:  ', self.percentDone
        print 'time left:     ', self.timeEst
        print 'download to:   ', self.downloadTo
        print 'download rate: ', self.downRate
        print 'upload rate:   ', self.upRate
        print 'share rating:  ', self.shareRating
        print 'seed status:   ', self.seedStatus
        print 'peer status:   ', self.peerStatus
        sys.stdout.flush()
        dpflag.set()

    def chooseFile(self, default, size, saveas, dir):
        self.file = '%s (%.1f MB)' % (default, float(size) / (1 << 20))
        if saveas != '':
            default = saveas
        self.downloadTo = os.path.abspath(default)
        return default

    def newpath(self, path):
        self.downloadTo = path


def run(params):
    h = HeadlessDisplayer()
    while 1:
        configdir = ConfigDir('downloadheadless')
        defaultsToIgnore = ['responsefile', 'url', 'priority']
        configdir.setDefaults(defaults, defaultsToIgnore)
        configdefaults = configdir.loadConfig()
        defaults.append(
            ('save_options', 0, 'whether to save the current options as the '
             'new default configuration (only for btdownloadheadless.py)'))
        try:
            config = parse_params(params, configdefaults)
        except ValueError as e:
            print 'error: {}\nrun with no args for parameter explanations' \
                ''.format(e)
            break
        if not config:
            print get_usage(defaults, 80, configdefaults)
            break
        if config['save_options']:
            configdir.saveConfig(config)
        configdir.deleteOldCacheData(config['expire_cache_data'])

        myid = createPeerID()
        random.seed(myid)

        doneflag = threading.Event()

        def disp_exception(text):
            print text
        rawserver = RawServer(
            doneflag, config['timeout_check_interval'], config['timeout'],
            ipv6_enable=config['ipv6_enabled'], failfunc=h.failed,
            errorfunc=disp_exception)
        upnp_type = UPnP_test(config['upnp_nat_access'])
        while True:
            try:
                listen_port = rawserver.find_and_bind(
                    config['minport'], config['maxport'], config['bind'],
                    ipv6_socket_style=config['ipv6_binds_v4'],
                    upnp=upnp_type, randomizer=config['random_port'])
                break
            except socket.error as e:
                if upnp_type and e == UPnP_ERROR:
                    print 'WARNING: COULD NOT FORWARD VIA UPnP'
                    upnp_type = 0
                    continue
                print "error: Couldn't listen - " + str(e)
                h.failed()
                return

        response = get_response(config['responsefile'], config['url'], h.error)
        if not response:
            break

        infohash = hashlib.sha1(bencode(response['info'])).digest()

        dow = BT1Download(
            h.display, h.finished, h.error, disp_exception, doneflag, config,
            response, infohash, myid, rawserver, listen_port, configdir)

        if not dow.saveAs(h.chooseFile, h.newpath):
            break

        if not dow.initFiles(old_style=True):
            break
        if not dow.startEngine():
            dow.shutdown()
            break
        dow.startRerequester()
        dow.autoStats()

        if not dow.am_I_finished():
            h.display(activity='connecting to peers')
        rawserver.listen_forever(dow.getPortHandler())
        h.display(activity='shutting down')
        dow.shutdown()
        break
    try:
        rawserver.shutdown()
    except Exception:
        pass
    if not h.done:
        h.failed()

if __name__ == '__main__':
    if sys.argv[1:] == ['--version']:
        print version
        sys.exit(0)

    if PROFILER:
        import profile
        import pstats
        p = profile.Profile()
        p.runcall(run, sys.argv[1:])
        log_fname = 'profile_data.' + time.strftime('%y%m%d%H%M%S') + '.txt'
        with open(log_fname, 'a') as log:
            normalstdout, sys.stdout = sys.stdout, log
            pstats.Stats(p).strip_dirs().sort_stats('time').print_stats()
            sys.stdout = normalstdout
    else:
        run(sys.argv[1:])
